subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestFramework.java

```diff
@@ -56,10 +56,7 @@ public WorkerTestClassProcessorFactory getProcessorFactory() {
     public Action<WorkerProcessBuilder> getWorkerConfigurationAction() {
         return new Action<WorkerProcessBuilder>() {
             public void execute(WorkerProcessBuilder workerProcessBuilder) {
-                workerProcessBuilder.sharedPackages("org.junit.platform.engine");
-                workerProcessBuilder.sharedPackages("org.junit.platform.commons");
-                workerProcessBuilder.sharedPackages("org.junit.platform.launcher.core");
-                workerProcessBuilder.sharedPackages("org.junit.platform.launcher");
+                workerProcessBuilder.sharedPackages("org.junit");
```

Comment:
This is changed because I misunderstood the meaning of `sharedPackages`. We need to share all `org.junit` classes (e.g. `RunWith` and `Enclosed`) for `JUnitPlatformTestFramework`.

Reply:
This shouldn't be necessary, we can test the name of the class instead of doing instanceof/equals. We shouldn't leak our JUnit version into the user code.

Reply:
`equals`/`instanceof` can't get `Runner` class from `RunWith`, we have to use reflection to invoke `RunWith.value()` if we want to make this happen:

```
        RunWith runWith = outermostClass.getAnnotation(RunWith.class);
        return runWith != null && Enclosed.class.equals(runWith.value());
```

In my understanding, `sharedPackages` means we should use jars provided by users first:

```
ExtClassLoader
            |
FilteringClassLoader <- this filters packages which aren't shared
            |
classloader with worker jars 
```

If users provide a specific version, `sharedPackages("org.junit")` allows our worker to use them. Is there any reason not to do like this?

Reply:
What makes you think we will use the user jars first? Classloading always works from the parent first. So if you share a package, then our version will be used.

Reply:
Well there must be some misunderstanding, let's try to explain (my understanding) in detail.

Our worker's classloader structure are like this:

```
ExtClassLoader
            |
FilteringClassLoader
            |
WorkerClassloader
```

Where `ExtClassloader` contains jars from `testRuntime` dependencies, let's say the user uses JUnit 4.11.

`FilteringClassLoader` determines which package can be shared, which can not. For JUnit, `org.junit` will be [shared](https://github.com/gradle/gradle/blob/2f94bfed05d248447439febefb45e841b7d73793/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestFramework.java#L57)

WorkerClassloader with worker jars (specified [here](https://github.com/gradle/gradle/blob/5675637494609449e322a90036c8b38a54bb6c22/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessor.java#L136)).

When worker process JVM needs a class, e.g. `org.junit.Test`, it would fetch from `ExtClassLoader`, which means user jars come first. Then `FilteringClassLoader` [checks if this class can be shared](https://github.com/gradle/gradle/blob/76054d7eb31bc7d8e06e960c4720fd22e8edb3b0/subprojects/base-services/src/main/java/org/gradle/internal/classloader/FilteringClassLoader.java#L91):

```
        try {
            return EXT_CLASS_LOADER.loadClass(name);
        } catch (ClassNotFoundException ignore) {
            // ignore
        }

        if (!classAllowed(name)) {
            throw new ClassNotFoundException(name + " not found.");
        }
```

Now it's `WorkerClassLoader`'s turn. We can do some experiment to verify "user jars come first when loading shared class":

Adds serveral lines in `JUnitTestClassProcessor`:

```
        File jar = ClasspathUtil.getClasspathForClass(Test.class);
        if (!jar.getAbsolutePath().contains(System.getProperty("junit.version"))) {
            throw new IllegalStateException("Test path: " + jar.getAbsolutePath() + "; junit.version: " + System.getProperty("junit.version"));
        }
```

Build a custom gradle distribution and run:

```
dependencies {
    testCompile 'junit:junit:4.11'
}

test {
    systemProperty('junit.version', '4.12')
}
```

This will throw exception, which proves "user jars come first when loading shared class".

Am I wrong anywhere?

Reply:
The reason why we need to share `org.junit` in both JUnit 4&5 is this line:

```
RunWith runWith = outermostClass.getAnnotation(RunWith.class);
return runWith != null && Enclosed.class.equals(runWith.value());
```

It works fine for JUnit 4, but if JUnit 5 Vintage engine tries to invoke this method, `RunWith.class` will be loaded by different classloaders (`ExtClassLoader` and `WorkerClassLoader` above) because `org.junit` aren't shared. Since shared package contains all descendant packages - i.e. if you share `org.junit`, you share `org.junit.*` as well - so `workerProcessBuilder.sharedPackages("org.junit");` is the only thing needed here.

Reply:
Thanks for trying this out - I had assumed that the ExtClassloader contains our infrastructure and the WorkerClassLoader contains the tests, but it seems to be the other way round. In that case sharing the package is indeed okay.